### ISSUE TITLE
Example for storing & fetching string/bytes ?

### ISSUE THREAD

**Issue by hazcod** [google/go-tpm#164]
Hi,
I'd like to use this library to fetch a master key out of the TPM chip.
This master key should only be available to my Go program.

Is there a best practice example I could use for this use case?
Altough this seems like a typical scenario, I am having issues grasping the API.
Thank you!

**Comment by twitchyliquid64** [comment]
> This master key should only be available to my Go program.

I don't think this is possible. The TPM communicates with the machine over a low-speed data bus, so any program on the host could send the right command to get the master key. The TPM has no way to know what program is firing off the commands under the hood.

Also, you probably want to use tpm2.Create to make a key & Load it, rather than assuming theres free NVRAM space. With Create, the TPM gives you an encrypted blob to hold onto (which is the key), and unwraps it into a key you can use when you give it back.

**Comment by hazcod** [comment]
@twitchyliquid64 Is this possible when shielded by filesystem permissions? (e.g. `chown root:root /dev/tpm`) What command are you talking about?

Doesn't `CreatePrimary(rw io.ReadWriter, owner tpmutil.Handle, sel PCRSelection, parentPassword, ownerPassword string, p Public) (tpmutil.Handle, crypto.PublicKey, error)` store an assymetric key, while I would be needing a symmetric key?

**Comment by twitchyliquid64** [comment]
> Is this possible when shielded by filesystem permissions? (e.g. chown root:root /dev/tpm)

Whether filesystem perms will be sufficient against your stated goal (_This master key should only be available to my Go program._) depends entirely on your threat model.

But if that was the case, why use the TPM at all? Why not just have the key in a file, and use permissions to control access?

> Doesn't CreatePrimary(rw io.ReadWriter, owner tpmutil.Handle, sel PCRSelection, parentPassword, ownerPassword string, p Public) (tpmutil.Handle, crypto.PublicKey, error) store an assymetric key, while I would be needing a symmetric key?

Correct, I assumed you needed an asymmetric key.

> Or would I need to follow something comparable to `tpm2-seal-unseal/` ...

If you want to store a secret on the TPM, then seal/unseal is the right way to go.

> where I would need to pass the following to fetch my secret

I think so, but would need to play around to be sure.

In almost all cases you will need an SRK (and any auth value necessary to use it), the ownership password (usually all zeros), the TPM to be owned, etc.

**Comment by hazcod** [comment]
The use case here is that a specific program should be fed the master key automatically, but limiting the attack surface as much as possible. It's a good point that once you are root, it's game over for both TPM & file-based restrictions. Are there programs that tackle this in some way?
I suppose the big plus is that if hard disks get 'lost' or stolen, the secret is still embedded on the motherboard/TPM board. Vice versa, if the motherboard gets stolen, they still need access to the disk for the owner/srk secrets + secret position value.

Am I correct in stating that using e.g. PCR 4 will allow linux disk encryption (LUKS) to use e.g. PCR 3?
(Seems trustedgrub uses PCR 0-13 https://github.com/fox-it/linux-luks-tpm-boot#install-trustedgrub2, so PCR 14-23 is still available)

What is the correct way to retrieve the SRK handle without sealing data first?
`tpm2.Load(rwc, srkHandle, srkPassword, publicArea, privateArea)`

**Comment by twitchyliquid64** [comment]
> Are there programs that tackle this in some way?

This is one of the main challenges in building secure systems imho - intra-system provenance & trust. Its common to dedicate a single physical machine to be the 'secrets management' or CA, and harden the network as its security boundary. Theres less variables there than trying to do it on the one system.

There are certainly ways to build a system where it is really hard to get around mandatory access controls, but even an ideal defense-in-depth system is still gonna have to face unpatched privescs, of which truckloads of undiscovered ones definitely exist in Linux.

**Comment by twitchyliquid64** [comment]
> I suppose the big plus is that if hard disks get 'lost' or stolen, the secret is still embedded on the motherboard/TPM board. Vice versa, if the motherboard gets stolen, they still need access to the disk for the owner/srk secrets + secret position value.

Yepp :) and if you seal the secret against a PCR, it gets even harder.

> Am I correct in stating that using separate PCRs will allow use of different technologies using the TPM, e.g. LUKS linux disk encryption?

I don't really understand the question. PCRs are not consumable resources - think of them as just a read-only bit of data, the length of a hash. The only way you can modify a PCR is by calling _extend_, which replaces the PCR with the hash of the current value + input.

Different PCRs are typically _extended_ into by the firmware / OS for different purposes. But you can use whatever PCR you like to seal a secret.

> What is the correct way to retrieve the SRK handle without sealing data first?

TPM 1.2 or 2.0?

On 2.0 its common to store it in a specific index, but if it doesnt exist you can create it with `CreatePrimary`. This works fine because primary keys like the SRK are deterministic.
